app/scripts/lib/metaRPCClient.js

```diff
@@ -0,0 +1,77 @@
+import { EventEmitter } from 'events';
+import { EthereumRpcError } from 'eth-rpc-errors';
+import createRandomId from '../../../shared/modules/random-id';
+
+class MetaRPCClient {
+  constructor(connectionStream) {
+    this.connectionStream = connectionStream;
+    this.notificationChannel = new EventEmitter();
+    this.requests = new Map();
+    this.connectionStream.on('data', this.handleResponse.bind(this));
+    this.connectionStream.on('end', this.close.bind(this));
+    // eslint-disable-next-line no-constructor-return
+    return new Proxy(this, {
+      get: (object, property) => {
+        if (object[property]) {
+          return object[property];
+        }
+        return (...p) => {
+          const cb = p[p.length - 1];
+          const params = p.slice(0, -1);
+          const id = createRandomId();
+
+          object.requests.set(id, cb);
+          object.connectionStream.write({
+            method: property,
+            params,
+            id,
+          });
+        };
+      },
+    });
+  }
+
+  onNotification(handler) {
+    this.notificationChannel.addListener('notification', (data) => {
+      handler(data);
+    });
+  }
+
+  close() {
+    this.notificationChannel.removeAllListeners();
+  }
+
+  handleResponse(data) {
+    const { id, result, error, method, params } = data;
+    const cb = this.requests.get(id);
+
+    if (method && params && id) {
+      // dont handle server-side to client-side requests
+      return;
+    }
+    if (method && params && !id) {
+      // handle servier-side to client-side notification
+      this.notificationChannel.emit('notification', data);
+      return;
+    }
+    if (!cb) {
+      // not found in request list
```

Comment:
Similarly, we might want to throw here. I'm guessing that it is possible we'll hit this case, if we every reply twice for any reason. But if we're doing that, it would be a bug that we'd want to track down.

Reply:
This could happen if you made 2 instance of this class. One of them would try to handle an id it doesn't care about.

Reply:
That would still be a bug, wouldn't it? At least I can't think of a case where this would be expected. Even if there were two instances, hopefully they'd be handling different sets of requests, i.e. we wouldn't have a request from one reply to the other.

Reply:
```
const client1 = metaRPCClientFactory(connectionStream);
client1.bar();
// random id: 123456

const client2 = metaRPCClientFactory(connectionStream);
client2.baz();
// random id: 54321
```

in this example, by calling `baz()` on client2 would still get a result over `connectionStream` triggering `onResponse` on `client1` which then would look for a callback (54321) it doesn't care about.

Reply:
Ohhhhhh I see, I get it now. Thanks, I didn't realize this would broadcast each reply to all clients.

Is that how our post message stream works? :thinking: